<div style="text-align: center; background-color: #750E21; font-family: 'Trebuchet MS', Arial, sans-serif; color: white; padding: 20px; font-size: 40px; font-weight: bold; border-radius: 0 0 0 0; box-shadow: 0px 6px 8px rgba(0, 0, 0, 0.2);">
  PROJECT: ... 📌
</div>

<div style="text-align: center; background-color: #0766AD; font-family: 'Trebuchet MS', Arial, sans-serif; color: white; padding: 20px; font-size: 40px; font-weight: bold; border-radius: 0 0 0 0; box-shadow: 0px 6px 8px rgba(0, 0, 0, 0.2);">
  Stage 3- MODELING 📌
</div>

## **IMPORT LIBRARY** 🎄

In [1]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split 
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

Đầu tiên, ta sẽ đọc tập dữ liệu đã được xử lý ở giai đoạn EDA.

In [2]:
df = pd.read_csv('../Data/2425_QC_EDA.csv')
df.head(10)

Area Income  Age  Daily Time Spent on Site  Daily Internet Usage  \
0     61833.90   35                     68.95                256.09   
1     68441.85   31                     80.23                193.77   
2     59785.94   26                     69.47                236.50   
3     54806.18   29                     74.15                245.89   
4     73889.99   35                     68.37                225.58   
5     59761.56   23                     59.99                226.74   
6     53852.85   33                     88.91                208.36   
7     24593.33   48                     66.00                131.76   
8     68862.00   30                     74.53                221.51   
9     55642.32   20                     69.88                183.82   

   Clicked on Ad  
0              0  
1              0  
2              0  
3              0  
4              0  
5              0  
6              0  
7              1  
8              0  
9              0

Trong giai đoạn EDA, ta đã kiểm tra được các thuộc tính có ảnh hưởng đến việc người dùng có nhấp vào quảng cáo hay không. Và đây là những thuộc tính mà chúng ta sẽ dùng để xây dựng Model:
- Area Income
- Age
- Daily Time Spent on Site
- Daily Internet Usage

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #FF90BC; padding: 20px; font-size: 28px; font-weight: bold; border-radius: 0 0 0 0">
  Splitting Dataset 🔥
</div>

Trước tiên, chúng ta cần phải chia ra 2 tập dữ liệu Train và Test để huấn luyện và thực hiện kiểm tra sau này.

In [73]:
random_state = 23

target = 'Clicked on Ad'

X, y = df.drop(target, axis=1), df[target]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=random_state)\


scaler = MinMaxScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

## **XÂY DỰNG MÔ HÌNH DỰ ĐOÁN** 📝

Bây giờ, ta sẽ xây dựng các mô hình để dự đoán kết quả người dùng có click vào quảng cáo hay không.

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  1. Logistic Regression 🔥
</div>

In [74]:
# Sử dụng Logistic Regression, hãy xây dựng mô hình dự đoán target dựa trên các biến đầu vào.
LR_model = LogisticRegression(random_state=random_state, solver='lbfgs', max_iter=500)
LR_model.fit(X_train, y_train)

y_pred = LR_model.predict(X_test)

LR_accuracy = accuracy_score(y_test, y_pred)

print(f'Accuracy: {LR_accuracy}')

Accuracy: 0.97


### **Nhận xét** 📈

Sau khi áp dụng đầy đủ các bước chuẩn bị và tiền xử lý dữ liệu cũng như EDA, mô hình Logistic Regression đã đạt được kết quả rất tốt với độ chính xác là 0.97.

- Việc làm sạch dữ liệu, xử lý các giá trị bị thiếu, và cân bằng phân phối các lớp đã đảm bảo rằng dữ liệu đầu vào không gây nhiễu hoặc thiên lệch cho mô hình.
- Loại bỏ các đặc trưng không liên quan và xử lý đa cộng tuyến đã giúp mô hình chỉ học từ các thông tin thực sự quan trọng, tối ưu hóa khả năng phân biệt.
- Sử dụng MinMaxScaler để chuẩn hóa dữ liệu đã giúp mô hình Logistic Regression hoạt động ổn định hơn, vì các đặc trưng đầu vào nằm trong cùng một phạm vi giá trị, loại bỏ ảnh hưởng của các giá trị lớn.
- Việc sử dụng train_test_split đảm bảo mô hình được đánh giá trên tập dữ liệu độc lập, giúp kết quả đáng tin cậy và phản ánh đúng khả năng tổng quát của mô hình.

Kết quả 0.97 cho thấy các bước xử lý đã góp phần quan trọng vào hiệu suất của mô hình, đồng thời khẳng định rằng Logistic Regression là một lựa chọn phù hợp cho bài toán với bộ dữ liệu này.

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  2. Naive Bayes 🔥
</div>

In [5]:
# Sử dụng Naive Bayes, hãy xây dựng mô hình dự đoán target dựa trên các biến đầu vào.
NB_model = GaussianNB()
NB_model.fit(X_train, y_train)

y_pred = NB_model.predict(X_test)

NB_accuracy = accuracy_score(y_test, y_pred)

print(f'Accuracy: {NB_accuracy}')

Accuracy: 0.97


### **Nhận xét** 📈
- Ta sử dụng Naive Bayes vì phụ thuộc giữa các cột dữ liệu là gần như là không có và các cột Area Income, Age, Daily Time Spent on Site, và Daily Internet Usage đều là dữ liệu liên tục
- Naive Bayes không yêu cầu lượng dữ liệu huấn luyện lớn để hoạt động tốt.
- Độ chính xác khi sử dụng Naive Bayes cho ra là 97%, khá cao cho một mô hình phân lớp cơ bản.
- Naive Bayes không yêu cầu xử lý phức tạp với dữ liệu (như tạo đặc trưng mới, tương tác giữa các đặc trưng), mà tập trung vào việc tính xác suất dựa trên các thuộc tính đã chọn.

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  3. Decision Tree 🔥
</div>

In [6]:
# Sử dụng Decision Tree, hãy xây dựng mô hình dự đoán target dựa trên các biến đầu vào.
DT_model = DecisionTreeClassifier(random_state=random_state)

DT_model.fit(X_train, y_train)

y_pred = DT_model.predict(X_test)

DT_accuracy = accuracy_score(y_test, y_pred)

print(f'Accuracy: {DT_accuracy}')
print("Confusion Matrix:") 
print(confusion_matrix(y_test, y_pred)) 
print("\nClassification Report:") 
print(classification_report(y_test, y_pred))

Accuracy: 0.95
Confusion Matrix:
[[91  7]
 [ 3 99]]

Classification Report:
              precision    recall  f1-score   support

           0       0.97      0.93      0.95        98
           1       0.93      0.97      0.95       102

    accuracy                           0.95       200
   macro avg       0.95      0.95      0.95       200
weighted avg       0.95      0.95      0.95       200



### **Nhận xét** 📈
- Mô hình có khả năng nhận biết với độ chính xác cao (0.95).
- Số lượng dự đoán sai (7 False Positives và 3 False Negative) là rất thấp, cho thấy mô hình có khả năng phân loại chính xác và ít nhầm lẫn, với độ nhạy của lớp "Nhấn vào Ad" cao hơn lớp "Không nhấn vào Ad".
- Khả năng phân loại chính xác lớp "Không nhấn vào Ad" cao hơn lớp "nhấn vào Ad".
- F1-score cao cho thấy sự cân bằng tốt giữa độ chính xác và độ nhạy.
- Việc sử dụng Decision Tree có thể giúp việc biểu diễn dữ liệu trực quan hơn khi có thể vẽ cây quyết định.
- Đồng thời cũng có thể sử dụng dữ liệu thô mà không cần chuẩn hóa.

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  4. Random Forest 🔥
</div>

In [7]:
# Sử dụng Random Forest, hãy xây dựng mô hình dự đoán target dựa trên các biến đầu vào.
RF_model = RandomForestClassifier(random_state=random_state)
RF_model.fit(X_train, y_train)

y_pred = RF_model.predict(X_test)

RF_accuracy = accuracy_score(y_test, y_pred)

print(f'Accuracy: {RF_accuracy}')

Accuracy: 0.97


### **Nhận xét** 📈
Thuật toán Random Forest (RF) là một mô hình học máy mạnh mẽ, đặc biệt hiệu quả với các bài toán phân loại và hồi quy. Dựa trên kết quả Accuracy = 0.97, đây là một kết quả rất tốt, cho thấy RF hoạt động hiệu quả với dữ liệu.

- Với độ chính xác 97%, mô hình Random Forest đã phân loại rất tốt, chứng tỏ nó có thể nắm bắt được mối quan hệ giữa các biến đầu vào và biến mục tiêu.

- Do Random Forest sử dụng nhiều cây quyết định (decision trees) và mỗi cây được huấn luyện trên một mẫu ngẫu nhiên, nó giảm thiểu nguy cơ overfitting so với việc sử dụng một cây quyết định duy nhất.

- Random Forest có thể mô hình hóa các mối quan hệ phức tạp giữa các biến mà không yêu cầu giả định tuyến tính.

- RF không bị ảnh hưởng nhiều bởi các giá trị ngoại lai hoặc dữ liệu nhiễu, nhờ vào sự tổng hợp kết quả từ nhiều cây.

## **ĐÁNH GIÁ CHẤT LƯỢNG MÔ HÌNH** 📝

Sau khi đã xây dựng các mô hình, ta sẽ thực hiện đánh giá các mô hình

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  1. Cross Validation với Precision 🔥
</div>

In [8]:
# Sử dụng dụng Cross Validation với recall để đánh giá các mô hình
LR_precision = cross_val_score(LR_model, X, y, cv=5, scoring='precision').mean()
NB_precision = cross_val_score(NB_model, X, y, cv=5, scoring='precision').mean()
DT_precision = cross_val_score(DT_model, X, y, cv=5, scoring='precision').mean()
RF_precision = cross_val_score(RF_model, X, y, cv=5, scoring='precision').mean()

print(f'Precision (Logistic Regression): {LR_precision:.4f}')
print(f'Precision (Naive Bayes): {NB_precision:.4f}')
print(f'Precision (Decision Tree): {DT_precision:.4f}')
print(f'Precision (Random Forest): {RF_precision:.4f}')

Precision (Logistic Regression): 0.9758
Precision (Naive Bayes): 0.9680
Precision (Decision Tree): 0.9273
Precision (Random Forest): 0.9622


### **Nhận xét** 📈

Trong bài toán dự đoán việc nhấp vào quảng cáo, với nhãn dương là hành động "nhấp vào quảng cáo", Precision là chỉ số quan trọng vì chúng ta muốn tối đa hóa độ chính xác khi dự đoán nhãn dương. Điều này giúp giảm thiểu các trường hợp dự đoán nhầm người không nhấp vào quảng cáo (nhãn âm) thành nhấp vào quảng cáo (nhãn dương), tránh gây lãng phí tài nguyên trong chiến dịch quảng cáo.

- Logistic Regression (Precision: 0.9758): Logistic Regression có Precision cao nhất, cho thấy mô hình này rất hiệu quả trong việc dự đoán chính xác người sẽ nhấp vào quảng cáo. Đây là lựa chọn hàng đầu nếu mục tiêu là tối ưu hóa các chiến dịch quảng cáo dựa trên dự đoán nhãn dương chính xác.

- Naive Bayes (Precision: 0.9680): Naive Bayes có Precision gần ngang với Logistic Regression, cũng là một lựa chọn tốt. Dù không mạnh bằng Logistic Regression, mô hình này vẫn đáng cân nhắc khi cần một giải pháp đơn giản và hiệu quả.

- Decision Tree (Precision: 0.9273): Precision thấp nhất của Decision Tree cho thấy mô hình này dễ nhầm lẫn hơn giữa nhãn dương và nhãn âm. Điều này có thể gây lãng phí chi phí quảng cáo vì dự đoán sai nhiều trường hợp không nhấp vào quảng cáo thành nhấp vào quảng cáo.

- Random Forest (Precision: 0.9622): Random Forest cải thiện đáng kể Precision so với Decision Tree, chứng tỏ mô hình này có khả năng khái quát tốt hơn nhờ tổng hợp nhiều cây quyết định. Đây là một lựa chọn phù hợp nếu cần sự cân bằng giữa hiệu suất cao và khả năng xử lý dữ liệu phức tạp.

Kết luận: Logistic Regression với Precision 0.9758 là lựa chọn tốt nhất cho bài toán này, vì nó dự đoán chính xác nhất hành động nhấp vào quảng cáo, đáp ứng được mục tiêu tối ưu hóa tài nguyên quảng cáo. Tuy nhiên, Random Forest và Naive Bayes cũng có thể được cân nhắc dựa trên các yếu tố khác như tốc độ huấn luyện, triển khai, và độ phức tạp của dữ liệu.

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  2. Cross Validation với Recall 🔥
</div>

In [9]:
# Sử dụng dụng Cross Validation với Recall để đánh giá các mô hình
LR_recall = cross_val_score(LR_model, X, y, cv=5, scoring='recall').mean()
NB_recall = cross_val_score(NB_model, X, y, cv=5, scoring='recall').mean()
DT_recall = cross_val_score(DT_model, X, y, cv=5, scoring='recall').mean()
RF_recall = cross_val_score(RF_model, X, y, cv=5, scoring='recall').mean()

print(f'Recall (Logistic Regression): {LR_recall:.4f}')
print(f'Recall (Naive Bayes): {NB_recall:.4f}')
print(f'Recall (Decision Tree): {DT_recall:.4f}')
print(f'Recall (Random Forest): {RF_recall:.4f}')

Recall (Logistic Regression): 0.9580
Recall (Naive Bayes): 0.9620
Recall (Decision Tree): 0.9440
Recall (Random Forest): 0.9580


### **Nhận xét** 📈
- ...

<div style="text-align: left; font-family: 'Trebuchet MS', Arial, sans-serif; color: #8C3333; padding: 20px; font-size: 25px; font-weight: bold; border-radius: 0 0 0 0; background-color: #AFC8AD">
  3. Cross Validation với F1 🔥
</div>

In [10]:
# # Sử dụng dụng Cross Validation với F1 để đánh giá các mô hình
LR_f1 = cross_val_score(LR_model, X, y, cv=5, scoring='f1').mean()
NB_f1 = cross_val_score(NB_model, X, y, cv=5, scoring='f1').mean()
DT_f1 = cross_val_score(DT_model, X, y, cv=5, scoring='f1').mean()
RF_f1 = cross_val_score(RF_model, X, y, cv=5, scoring='f1').mean()

print(f'F1 (Logistic Regression): {LR_f1:.4f}')
print(f'F1 (Naive Bayes): {NB_f1:.4f}')
print(f'F1 (Decision Tree): {DT_f1:.4f}')
print(f'F1 (Random Forest): {RF_f1:.4f}')

F1 (Logistic Regression): 0.9666
F1 (Naive Bayes): 0.9648
F1 (Decision Tree): 0.9347
F1 (Random Forest): 0.9599


### **Nhận xét** 📈
- ...

## **KẾT LUẬN** 📝